### DISCUSSION TITLE
Simple SSO

### DISCUSSION THREAD

**Discussion by Kixunil** [btcpayserver/btcpayserver#1973] [Ideas / Feature Requests]
**Is your feature request related to a problem? Please describe.**
I'd like to provide the users of [my repository](https://deb.ln-ask.me) with a way to authenticate only once and then access any application without further authentication. This improves security as the users don't have to remember/store multiple passwords and makes it easier to use applications.

I'm currently working on a simple tool that will allow the users to access all the applications after logging in. It already works with Thunderhub and RTL. I'd love it to work with BTCPayServer as well.

**Describe the solution you'd like**
I'd like to have an option to somehow retrieve a secret token for a user that can be appended to the URL. e.g. `/btcpay?token=deadbeef`. BTCPay should identify the user and log him in as if he entered e-mail and password on the login page.

Further, I'd love it to work such that retrieving a token for non-existing user would create a new user account.
This should also work for the admin account that's accessed during the initialization. Or even better, have an option to create such account immediately after start so that only specific user can access it.

Finally it'd be really nice if  user names could be arbitrary, not just e-mail address. E-mail addresses are scary/suspicious for all privacy-conscious people.

What I'm imagining is to have new options `managementTokenPath` and `createAdminIfNotExist` in the configuration **file**. A random cookie should be written to file referenced by `managementTokenPath`. A helper program could then use this token to call `somepath/token/$USERNAME?mgmtToken=...` to retrieve the token for user `$USERNAME`. It'd then pass this token to the user in a link, so when a user clicks this link he's automatically logged in without further action.

**Describe alternatives you've considered**
I could write an external program that would access BTCPay and manage the users manually. Then it'd generate tokens and URLs with them pointing to a special btcpaysubpath, which would be shadowed by nginx black magic to secretly log the user in and obtain appropriate cookies. Why this would work in theory, it'd be insanely complicated and fragile. Any change of BTCPay API could break it. I don't think it'd be a good idea.
I could also make an exception and force users to log into BTCPay manually. This could confuse them if they are used to not having to log in twice.

**Additional context**
If there's a good reason to use other methods to transfer the token to BTCPay (such as POST request) I wouldn't be opposed to it, but URL is the simplest since other projects use it too. The tool I'm building is called [`selfhost-dashboard`](https://github.com/Kixunil/selfhost-dashboard) and while not production-ready it already does something interesting.

**Comment by Kukks** [comment]
This is already possible using Greenfield: https://docs.btcpayserver.org/API/Greenfield/v1/#tag/Authorization/paths/~1api-keys~1authorize/get

**Comment by Kukks** [comment]
@dennisreimann and I worked on a proof of concept of SSO from an external application with a btcpay instance.
Maybe you can share some of the code @dennisreimann?

**Comment by Kixunil** [comment]
That seems to be doing something different. My understanding is that it enables people to generate API keys for other applications to use the **API**. I want to allow people to use the **UI** of BTCPay without having to type the password. My application is system-level, so it's considered trusted and may have all permissions.

**Comment by Kukks** [comment]
> This is already possible using Greenfield: https://docs.btcpayserver.org/API/Greenfield/v1/#tag/Authorization/paths/~1api-keys~1authorize/get

The flow here is:
* generate a specific authorize url with a specific appIdentifier and a redirect url
* redirect user to it
* user logs in/registers
* user approves
* user is redirected back to your application with an api key( you can specify which permissions you need)

**Comment by Kukks** [comment]
Ah, you want the opposite then.

**Comment by Kixunil** [comment]
Yes, I want other application to manage the accounts and enable users to access BTCPay without further manual authentication.

**Comment by Kixunil** [comment]
I offer a bounty of 100k sats to the fist person who securely implements SSO as described above with the ability to use arbitrary username (willing to discuss other options if this is difficult). Each of first three reviewers will get 10k sats. This offer expires in one month when I may or may not prolong it or change the amounts. The bounty will be paid after merge.

I offer 20k more for the ability to have admin account created on the first start of BTCPayServer (settable in config **file**) which avoids security risk.

I will do my best to be available for answering technical details and will myself review security properties of the implementation.

**Reply by Kukks** [reply]
Ok so technicalities here:
* Possibly introduce this as a plugin
* Introduce a config item: SSOCookieFilePath
* On start or cookie usage: generate a file with a secret inside a file at `SSOCookieFilePath`
* Create an HTTP POST endpoint that accepts a body:
```json
{
  "cookieToken": "contents of the file generated at SSOCookieFilePath here",
  "username" : "username/email of the user to log in",
  "newUserPassword": "if the user doesnt exist, this will be the password"
}
```
* If cookieToken provided is correct, 
  * regenerate the cookie file.
  * If user exists
    * log in user
  * else
    *  if `newUserPassword` is set, create the user with the specified password
    *  log in user

The username != email is a separate feature
The auto admin account via config is a separate feature (which can be built on the original proposed feature)

**Reply by Kixunil** [reply]
By "log in user" you mean return token for the specific user, right? `cookieToken` must not be known to browser - otherwise any user can login as any other.

Also there's some additional thing you need to get right but I have a good reason to talk about it privately. Is Keybase chat OK for you?

> The username != email is a separate feature

Do you agree to splitting 20k for this feature and 80k for SSO?

**Reply by Kukks** [reply]
Not doing this for the bounty honestly, I jut wanted to chime in and give specifics on how implementation could look like in case someone wanted to implement it,

Based on what you said, there's a few additional steps: 

* Possibly introduce this as a plugin
* Introduce a config item: SSOCookieFilePath
* On start or cookie usage: generate a file with a secret inside a file at `SSOCookieFilePath`
* Create an HTTP POST endpoint that accepts a body:
```json
{
  "cookieToken": "contents of the file generated at SSOCookieFilePath here",
  "username" : "username/email of the user to log in",
  "newUserPassword": "if the user doesnt exist, this will be the password"
}
```
* If cookieToken provided is correct, 
  * regenerate the cookie file.
  * If user exists
    * create a dedicated token that a user can use to log in with no credentials (expires in ~5mins)
  * else
    *  if `newUserPassword` is set, create the user with the specified password
    *  create a dedicated token that a user can use to log in with no credentials (expires in ~5mins)
* Create another HTTP Endpoint that takes generated tokens from the previous endpoint. Upon token submission log user in on browser and consume the token.

**Reply by Kixunil** [reply]
Yes, that looks correct.

**Reply by Kixunil** [reply]
I was thinking about it a bit more and I have an interesting alternative: instead of querying BTCPayServer for the temporary user token, generate it as `HMAC(monotonic_time || user_name, auth_token) || monotonic_time || user_name` and send straight to the user, which uses it for logging in. The token expires after one minute and there's no explicit invalidation. `auth_token` refers to the contents of `SSOCookieFilePath`.

This avoids roundtrip and probably makes the implementation simpler without causing security issues.

What do you think @Kukks?

**Comment by Kukks** [comment]
For the upcoming FIDO2 support, there may be a requirement of a feature that overlaps slightly. Since FIDO2 allows you to use your platform as the security key itself, there may be situations where a user can only authenticate by using one specific machine. We will most likely need to provide a feature that allows a user to "log in using another machine", perhaps by generating a link (and QR) on the authenticated machine that other machines can use to log in. 

We can probably expose this functionality over the API. Then a small application can be developed to generate these links/create users and a user is redirected to the URL for an SSO login.
